[DR-3037] Fix NPE when file mimetype is not specified

samanehsan opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3037

In addition to handling the NPE, this PR also converts the `TableFileDaoTest` class to use JUnit5.

samanehsan committed:
Fix NPE on mimeType

samanehsan committed:
Add unit test

samanehsan committed:
Convert TableFileDaoTest to JUnit5

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableFileDaoTest.java:
```diff
@@ -1,105 +1,119 @@
-    assertEquals("The same object is returned", files.get(0), expectedFiles.get(0));
+    @Mock private TableServiceClient tableServiceClient;
+    @Mock private TableClient tableClient;
+    @Autowired private TableFileDao dao;
```
In general you only want to use `@Autowired` in a test that's testing MVC or database operations. `TableFileDao` doesn't use either MVC or database operations so you don't need to autowire it, you can instead create a copy for testing using `new()`. Once you've done that, you can remove these annotations from the outer class:
```
@SpringBootTest
@AutoConfigureMockMvc
@EmbeddedDatabaseTest
```

```suggestion
    private final TableFileDao dao = new TableFileDao(mock(ExecutorService.class));
```

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableFileDaoTest.java:
```diff
@@ -1,105 +1,119 @@
+      List<FireStoreFile> expectedFiles = List.of(FireStoreFile.fromTableEntity(entity));
+      List<FireStoreFile> files =
+          dao.batchRetrieveFileMetadata(tableServiceClient, DATASET_ID, directoryEntries);
+      assertEquals(
+          files.size(), expectedFiles.size(), "A file record is found for each directory entry");
+      assertEquals(files.get(0), expectedFiles.get(0), "The same object is returned");
```
I would use `assertThat()` here instead to check both of these things at once:
```suggestion
      assertThat(files, is(expectedFiles));
```

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableFileDaoTest.java:
```diff
@@ -1,105 +1,119 @@
-  public void testRetrieveFileMetadata() {
-    FireStoreFile fileMetadata = dao.retrieveFileMetadata(tableServiceClient, DATASET_ID, FILE_ID);
+  void testFromTableEntityMissingMimeType() {
+    TableEntity entity = createTableEntity(false);
```
Maybe this should be in its own class, e.g., `FireStoreFileTest`?

samanehsan replied:
Good point! Then I can get rid of the nested class

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreFile.java:
```diff
@@ -198,7 +198,7 @@ public int hashCode() {
     return new FireStoreFile()
         .fileId(entity.getProperty(FILE_ID_FIELD_NAME).toString())
-        .mimeType(entity.getProperty(MIME_TYPE_FIELD_NAME).toString())
+        .mimeType((String) entity.getProperty(MIME_TYPE_FIELD_NAME))
```
Note that this is different if `entity.getProperty(MIME_TYPE_FIELD_NAME)` returns something other than `String`.

Also I see `toString()` is used on other fields here. Is it specified as part of the azure APIs which fields can and can't be `null`?

samanehsan replied:
I think the other fields should always be defined, but I can change those so that they are cast to `String` too.

pshapiro4broad replied:
Casting produces different behavior than `toString()` so I'd be careful. E.g. if a field is really `Integer` or `DateTime` then casting to `String` will result in a thrown `ClassCastException` at runtime, whereas calling `toString()` would not.

samanehsan replied:
Also the mimetype comes from `fileLoadModel.getMimeType()` so it will always be a string!

samanehsan replied:
The other properties are also strings, so the casting should be safe to do:
- https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzureFileStep.java#L50
- https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileFileStep.java#L42

pshapiro4broad reviewed (approved):
Looks OK, some possible test improvements

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableFileDaoTest.java:
```diff
@@ -1,105 +1,119 @@
-@ActiveProfiles({"google", "unittest"})
-@Category(Unit.class)
+@ExtendWith(MockitoExtension.class)
+@Tag("bio.terra.common.category.Unit")
```
Now that we've been migrating to `@Tag` we've been using a hardcoded string constant for the argument. I think we should instead define a constant for this and use that instead, to make it easier to find references to it and avoid possible typos. If a tag name is misspelled then a the test won't be run but no error will be generated.

E.g. something like:
```
package bio.terra.common.category;

public interface Unit {
  String TAG = "bio.terra.common.category.Unit";
}
```

and then
```
@Tag(Unit.TAG)
```

samanehsan committed:
Update tests

samanehsan committed:
Cast string fields

samanehsan committed:
Update com.jfrog.artifactory version

The previous version was using org.jfrog.buildinfo:build-info-extractor:2.15.0
which no longer exists.
